File: Number System Converter/BDConverter.java
import java.util.Scanner;

public class BDConverter {
	// Function interface for lambda expression.
	interface FuncInterface1 {
		void AbstractFunc1();
	}

	interface FuncInterface2 {
		String AbstractFunc2(String s);
	}


	static void ShowB2D(String bin) {
		int decimal = 0;
		for (int i = 0; i < bin.length(); i++) {
			decimal += Integer.parseInt(bin.substring(i,i+1)) * Math.pow(2, bin.length() - i - 1);
		}
		System.out.println("BIN " + bin + " = " + decimal + " DEC");
	}

	static void ShowD2B(String number) {
		String binary = "";
		int decimal = Integer.parseInt(number);
		while (decimal != 0) {
			binary = Integer.toString(decimal % 2) + binary;
			decimal /= 2;
		}
		System.out.println("DEC " + number + " = " + binary + " BIN");
	}


	static String B2D(String bin) {
		int decimal = 0;
		for (int i = 0; i < bin.length(); i++) {
			decimal += Integer.parseInt(bin.substring(i,i+1)) * Math.pow(2, bin.length() - i - 1);
		}
		return Integer.toString(decimal);
	}

	static String D2B(String number) {
		String binary = "";
		int decimal = Integer.parseInt(number);
		while (decimal != 0) {
			binary = Integer.toString(decimal % 2) + binary;
			decimal /= 2;
		}
		return binary;
	}

	public static void main(String[] args) {
		Scanner input = new Scanner(System.in);
		System.out.println("What number do you want to convert? ");
		String get = input.next();

		boolean decimal = false;
		for (int i = 0; i < get.length(); i++) {
			if (Integer.parseInt(get.substring(i,i+1)) > 1) {
				decimal = true;
				break;
			}
		}

		if (decimal) {
			System.out.println();
			System.out.println("Using lambda without parameter:");
			FuncInterface1 f = decimal ? () -> ShowD2B(get) : () -> ShowB2D(get);
			f.AbstractFunc1();

			System.out.println();
			System.out.println("Using lambda with parameter:");
			FuncInterface2 f2 = decimal ? str -> D2B(get) : str -> B2D(get);
			System.out.println("DEC " + get + " = " + f2.AbstractFunc2(get) + " BIN");
		} else {
			System.out.println();
			System.out.println("Using lambda without parameter:");
			FuncInterface1 f = () -> ShowD2B(get);
			f.AbstractFunc1();
			f = () -> ShowB2D(get);
			f.AbstractFunc1();
			

			System.out.println();
			System.out.println("Using lambda with parameter:");
			FuncInterface2 f2 = str -> D2B(get);
			System.out.println("DEC " + get + " = " + f2.AbstractFunc2(get) + " BIN");
			f2 = str -> B2D(get);
			System.out.println("BIN " + get + " = " + f2.AbstractFunc2(get) + " DEC");
		}
	}
}

File: Number System Converter/c/src/bdconverter.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "bdconverter.h"

char *D2B(char *number, char *binary, int length) {
	int capacity = length;
	int size = 0;
	int start = 0;
	int decimal = 0;
	sscanf(number, "%d", &decimal);
	binary = (char*)malloc(sizeof(char) * (capacity+1));
	memset(binary, '0', capacity+1);
	binary[length--] = '\0';

	while (decimal > 0) {
		binary[length--] = decimal % 2 + '0';
		decimal /= 2;
		size++;

		if (size >= capacity) {
			capacity *= 2;
			char *temp = (char*)malloc(sizeof(char) * (capacity+1));
			memset(temp, ' ', capacity+1);
			start = size;

			for (int i = capacity; i >= 0; i--) {
				if (i >= start) {
					temp[i] = binary[i-start];
				} else {
					for (int j = start-1; j > start - start%4; j--) {
						temp[j] = '0';
					}
					break;
				}
			}

			free(binary);
			binary = temp;
			length = size-1;
		}
	}

	if (binary[0] == ' ') {
		char *temp = (char *)malloc(sizeof(char) * (size + 1));
		memset(temp, 0, size+1);

		for (int i = 0; i < size; i++) {
			temp[i] = binary[i+capacity - size];
//			printf("size=%d temp[%d]=%c\n", size, i, temp[i]); 
//			printf("binary[%d]=%c\n", i+start, binary[i+capacity - size]); 
		}

		free(binary);
		binary = temp;
	}

	return binary;
}

void ShowD2B(char *number) {
	int length = 4;
	char *binary;

	printf("DEC %s = %s\n", number, D2B(number, binary, length));
	free(binary);
}

/*
void ShowD2B(char *number) {
	int capacity = 4, size = 0;
	int length = capacity;
	char *binary = (char*)malloc(sizeof(char) * (capacity+1));
	memset(binary, '0', capacity+1);

	int decimal = 0;
	sscanf(number, "%d", &decimal);
	binary[length--] = '\0';

	while (decimal > 0) {
		int space = capacity/4-1;
		if ((size+1)%5 == 0) {
			binary[length--] = ',';
		}
		binary[length--] = decimal % 2 + '0';
		decimal /= 2;
		size++;
//printf("binary[%d]=%c\n", length+1, binary[length+1]);
		if (size >= capacity) {
			capacity *= 2;
			char *temp = (char*)malloc(sizeof(char) * (capacity+1 + space));
			int start = size + space;

			for (int i = 0; i < capacity + space; i++) {
				if (i >= start) {
					temp[i] = binary[i-start];
				} else {
					temp[i] = '0';
				}
			}

			free(binary);
			binary = temp;
			length = size-1;
			printf("temp: %s\n", temp);
		}
//		printf("size=%d decimal=%d %s\n", size, decimal, binary);
	}

	printf("DEC %s = %s\n", number, binary);
	free(binary);
}
*/
void ShowB2D(char *binary) {
}

